# **IMPORT LIBRARY**

In [ ]:
import requests
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# **DATASET**



*Dataset* dapat diakses pada tautan berikut ini.

*   https://www.kaggle.com/datasets/irkaal/foodcom-recipes-and-reviews/data?select=reviews.csv

Kami juga meng-*upload* pada *drive* berikut ini.

*   https://drive.google.com/drive/folders/1mSpVckVreHO30nVf-UVN9yATu6pu5MZF?usp=sharing

*Dataset Reviews*

In [ ]:
# Load dataset

reviews = pd.read_csv("/content/drive/MyDrive/MK SISTEM REKOMENDASI/reviews.csv")
reviews

ReviewId  RecipeId    AuthorId        AuthorName  Rating  \
0               2       992        2008         gayg msft       5   
1               7      4384        1634     Bill Hilbrich       4   
2               9      4523        2046  Gay Gilmore ckpt       2   
3              13      7435        1773     Malarkey Test       5   
4              14        44        2085        Tony Small       5   
...           ...       ...         ...               ...     ...   
1401977   2090339    139499  2002080368       terrylbiggs       2   
1401978   2090340    148484    41805321        rogerberry       5   
1401979   2090341    264191  2002901848    Reiketsukan 6.       5   
1401980   2090345    411791  2002901938            Sue M.       5   
1401981   2090347    352686  2002812127        dede25cory       5   

                                                    Review  \
0             better than any you can get at a restaurant!   
1        I cut back on the mayo, and made up the differ...   
2        i think i did something wrong because i could ...   
3        easily the best i have ever had.  juicy flavor...   
4                                       An excellent dish.   
...                                                    ...   
1401977  I was disappointed. I couldn't wait to make th...   
1401978  Nothing to drain. And I don’t heat up the liqu...   
1401979  Good base recipe for someone to start with. I ...   
1401980  Thank you so much for this amazing recipe! I l...   
1401981  I can't say enough about this recipe.... the b...   

                DateSubmitted          DateModified  
0        2000-01-25T21:44:00Z  2000-01-25T21:44:00Z  
1        2001-10-17T16:49:59Z  2001-10-17T16:49:59Z  
2        2000-02-25T09:00:00Z  2000-02-25T09:00:00Z  
3        2000-03-13T21:15:00Z  2000-03-13T21:15:00Z  
4        2000-03-28T12:51:00Z  2000-03-28T12:51:00Z  
...                       ...                   ...  
1401977  2020-12-27T23:57:54Z  2020-12-27T23:57:54Z  
1401978  2020-12-28T00:44:42Z  2020-12-28T00:44:42Z  
1401979  2020-12-28T01:04:43Z  2020-12-28T01:04:43Z  
1401980  2020-12-28T03:07:10Z  2020-12-28T03:07:10Z  
1401981  2020-12-28T03:23:48Z  2020-12-28T03:23:48Z  

[1401982 rows x 8 columns]

*Dataset Recipes*

In [ ]:
# Load dataset

recipes = pd.read_csv("/content/drive/MyDrive/MK SISTEM REKOMENDASI/recipes.csv")
recipes

RecipeId                                              Name  \
0             38                 Low-Fat Berry Blue Frozen Dessert   
1             39                                           Biryani   
2             40                                     Best Lemonade   
3             41                    Carina's Tofu-Vegetable Kebabs   
4             42                                      Cabbage Soup   
...          ...                                               ...   
522512    541379                    Meg's Fresh Ginger Gingerbread   
522513    541380  Roast Prime Rib au Poivre with Mixed Peppercorns   
522514    541381                             Kirshwasser Ice Cream   
522515    541382          Quick & Easy Asian Cucumber Salmon Rolls   
522516    541383                           Spicy Baked Scotch Eggs   

          AuthorId      AuthorName CookTime PrepTime TotalTime  \
0             1533          Dancer    PT24H    PT45M  PT24H45M   
1             1567        elly9812    PT25M     PT4H   PT4H25M   
2             1566  Stephen Little     PT5M    PT30M     PT35M   
3             1586         Cyclopz    PT20M    PT24H  PT24H20M   
4             1538       Duckie067    PT30M    PT20M     PT50M   
...            ...             ...      ...      ...       ...   
522512  2002090414           rdsxc    PT35M     PT1H   PT1H35M   
522513      211566    Denver cooks     PT3H    PT30M   PT3H30M   
522514  2001131545     Jonathan F.     PT3H     PT1H      PT4H   
522515  2001004241       CLUBFOODY      NaN    PT15M     PT15M   
522516      188099    Chef LauraMD    PT25M    PT15M     PT40M   

               DatePublished  \
0       1999-08-09T21:46:00Z   
1       1999-08-29T13:12:00Z   
2       1999-09-05T19:52:00Z   
3       1999-09-03T14:54:00Z   
4       1999-09-19T06:19:00Z   
...                      ...   
522512  2020-12-22T15:27:00Z   
522513  2020-12-22T15:32:00Z   
522514  2020-12-22T15:33:00Z   
522515  2020-12-22T22:11:00Z   
522516  2020-12-22T22:12:00Z   

                                              Description  \
0       Make and share this Low-Fat Berry Blue Frozen ...   
1       Make and share this Biryani recipe from Food.com.   
2       This is from one of my  first Good House Keepi...   
3       This dish is best prepared a day in advance to...   
4       Make and share this Cabbage Soup recipe from F...   
...                                                   ...   
522512  Make and share this Meg's Fresh Ginger Gingerb...   
522513  White, black, green, and pink peppercorns add ...   
522514  Make and share this Kirshwasser Ice Cream reci...   
522515  Extremely quick and easy to make, these are gr...   
522516  Great way to have hard boiled eggs and crispy,...   

                                                   Images  ...  \
0       c("https://img.sndimg.com/food/image/upload/w_...  ...   
1       c("https://img.sndimg.com/food/image/upload/w_...  ...   
2       c("https://img.sndimg.com/food/image/upload/w_...  ...   
3       c("https://img.sndimg.com/food/image/upload/w_...  ...   
4       "https://img.sndimg.com/food/image/upload/w_55...  ...   
...                                                   ...  ...   
522512                                       character(0)  ...   
522513  "https://img.sndimg.com/food/image/upload/w_55...  ...   
522514                                       character(0)  ...   
522515  "https://img.sndimg.com/food/image/upload/w_55...  ...   
522516                                       character(0)  ...   

       SaturatedFatContent CholesterolContent SodiumContent  \
0                      1.3                8.0          29.8   
1                     16.6              372.8         368.4   
2                      0.0                0.0           1.8   
3                      3.8                0.0        1558.6   
4                      0.1                0.0         959.3   
...                    ...                ...           ...   
522512                

# **PRE-PROCESSING**

***FILTER* KOLOM *DATE SUBMITTED***

Melakukan pemisahan (*filter*) berdasarkan waktu rentang 26-01-2000 s.d. 31-12-2001 pada kolom *Date Submitted*.

In [ ]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1401982 entries, 0 to 1401981
Data columns (total 8 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   ReviewId       1401982 non-null  int64 
 1   RecipeId       1401982 non-null  int64 
 2   AuthorId       1401982 non-null  int64 
 3   AuthorName     1401982 non-null  object
 4   Rating         1401982 non-null  int64 
 5   Review         1401768 non-null  object
 6   DateSubmitted  1401982 non-null  object
 7   DateModified   1401982 non-null  object
dtypes: int64(4), object(4)
memory usage: 85.6+ MB


In [ ]:
# Mengubah kolom DateSubmitted menjadi tipe data datetime agar dapat melakukan filter berdasarkan waktu
reviews['DateSubmitted'] = pd.to_datetime(reviews['DateSubmitted'])
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1401982 entries, 0 to 1401981
Data columns (total 8 columns):
 #   Column         Non-Null Count    Dtype              
---  ------         --------------    -----              
 0   ReviewId       1401982 non-null  int64              
 1   RecipeId       1401982 non-null  int64              
 2   AuthorId       1401982 non-null  int64              
 3   AuthorName     1401982 non-null  object             
 4   Rating         1401982 non-null  int64              
 5   Review         1401768 non-null  object             
 6   DateSubmitted  1401982 non-null  datetime64[ns, UTC]
 7   DateModified   1401982 non-null  object             
dtypes: datetime64[ns, UTC](1), int64(4), object(3)
memory usage: 85.6+ MB


In [ ]:
# Melakukan filter
start_date = '2000-01-26'
end_date = '2001-12-31'
reviews = reviews[(reviews['DateSubmitted'] >= start_date) & (reviews['DateSubmitted'] <= end_date)]

reviews

ReviewId  RecipeId  AuthorId        AuthorName  Rating  \
1               7      4384      1634     Bill Hilbrich       4   
2               9      4523      2046  Gay Gilmore ckpt       2   
3              13      7435      1773     Malarkey Test       5   
4              14        44      2085        Tony Small       5   
5              17      5221      2046  Gay Gilmore ckpt       4   
...           ...       ...       ...               ...     ...   
1244723   1429155     10125     21656            Carol        5   
1244727   1429159     10125     16827       Ellen Brody       5   
1244731   1429163     10125      5076   selvi nadarajah       5   
1244738   1429170     10125     21267    Usako Yamakawa       2   
1244739   1429171     10125     26517             Ann S       5   

                                                    Review  \
1        I cut back on the mayo, and made up the differ...   
2        i think i did something wrong because i could ...   
3        easily the best i have ever had.  juicy flavor...   
4                                       An excellent dish.   
5                   love it, but without the bean sprouts.   
...                                                    ...   
1244723  WOW! These cookies are without a doubt the bes...   
1244727  These were really easy to make and are very so...   
1244731  thanks Dawn,\r\nthis was superb.I made it for ...   
1244738  Maybe I did something wrong, or maybe I was ex...   
1244739                                        Wonderful!!   

                    DateSubmitted          DateModified  
1       2001-10-17 16:49:59+00:00  2001-10-17T16:49:59Z  
2       2000-02-25 09:00:00+00:00  2000-02-25T09:00:00Z  
3       2000-03-13 21:15:00+00:00  2000-03-13T21:15:00Z  
4       2000-03-28 12:51:00+00:00  2000-03-28T12:51:00Z  
5       2000-05-08 11:08:00+00:00  2000-05-08T11:08:00Z  
...                           ...                   ...  
1244723 2001-11-22 10:48:19+00:00  2001-11-22T10:48:19Z  
1244727 2001-12-08 05:25:34+00:00  2001-12-08T05:25:34Z  
1244731 2001-11-14 09:37:41+00:00  2001-11-14T09:37:41Z  
1244738 2001-12-22 16:17:03+00:00  2001-12-22T16:17:03Z  
1244739 2001-12-18 10:50:55+00:00  2001-12-18T10:50:55Z  

[4010 rows x 8 columns]

**MENGAMBIL KOLOM YANG DIGUNAKAN**

Menghapus Kolom pada Dataset yang tidak diperlukan. Kolom yang akan digunakan adalah:


*   RecipeId
*   AuthorId
*   Rating

In [ ]:
# Membuat DataFrame
df = pd.DataFrame(reviews, columns=["AuthorId", "RecipeId", "Rating"])
df

AuthorId  RecipeId  Rating
1            1634      4384       4
2            2046      4523       2
3            1773      7435       5
4            2085        44       5
5            2046      5221       4
...           ...       ...     ...
1244723     21656     10125       5
1244727     16827     10125       5
1244731      5076     10125       5
1244738     21267     10125       2
1244739     26517     10125       5

[4010 rows x 3 columns]

In [ ]:
df.to_csv("/content/drive/MyDrive/MK SISTEM REKOMENDASI/Dataset Final.csv")

# ***SURPRISE LAB***



*Surprise* adalah *library* Python untuk membangun dan menganalisis sistem rekomendasi yang berkaitan dengan data *rating* eksplisit. *Surprise* dirancang dengan tujuan memberikan pengguna kendali yang sempurna atas eksperimen yang dilakukan, mengatasi masalah penanganan *dataset*, serta menyediakan berbagai algoritma prediksi siap pakai dan evaluasi untuk mengukur kinerja algoritma.

In [ ]:
pip install scikit-surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 6.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3163343 sha256=a7188f228a2817a8292295ad03de364c114714e7d121869dc51362d2014e3597
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [ ]:
# Import library Surprise
from surprise import Dataset, NormalPredictor, Reader
from surprise.model_selection import cross_validate, KFold
from surprise import SVD, accuracy
from sklearn.model_selection import train_test_split, GridSearchCV

In [ ]:
# Reader diperlukan dengan parameter rating_scale
reader = Reader(rating_scale=(1, 5))

# Kolom harus sesuai dengan user, item, dan rating
data = Dataset.load_from_df(df[["AuthorId", "RecipeId", "Rating"]], reader)

# **ALGORITMA SVD**



Algoritma *Singular Value Decomposition* (SVD) adalah teknik penguraian matriks yang digunakan dalam sistem rekomendasi untuk mengurai preferensi pengguna terhadap *item* ke dalam matriks pengguna, matriks *item*, dan matriks faktor tersembunyi. SVD memungkinkan sistem rekomendasi untuk membuat rekomendasi *item* berdasarkan faktor-faktor tersembunyi yang terkait dengan preferensi pengguna.

In [ ]:
# Cross-validation iterator
iterator = KFold(n_splits=3)

In [ ]:
# Menggunakan algoritma SVD
algo = SVD()

***Split Data***

In [ ]:
for trainset, testset in iterator.split(data):

    # train test
    algo.fit(trainset)
    predictions = algo.test(testset)

    # RMSE
    accuracy.rmse(predictions, verbose=True)

RMSE: 1.5167
RMSE: 1.5043
RMSE: 1.4792


# ***CROSSFOLD VALIDATION***



*Crossfold validation* adalah teknik evaluasi model yang digunakan untuk mengukur kinerja model dengan membagi *dataset* menjadi beberapa *fold*, melatih model, dan  mengukur sejauh mana model SVD secara konsisten memberikan rekomendasi yang baik pada berbagai bagian *dataset*. *Crossfold validation* membantu memastikan bahwa model tidak hanya bekerja dengan baik pada satu bagian data tertentu, tetapi juga pada seluruh *dataset*, meningkatkan akurasi dan dapat mengidentifikasi apakah model rentan terhadap *overfitting*.







***K = 2***

In [ ]:
results2 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=2, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 2 split(s).

                  Fold 1  Fold 2  Mean    Std     
MSE (testset)     2.3550  2.1938  2.2744  0.0806  
RMSE (testset)    1.5346  1.4811  1.5079  0.0267  
Fit time          0.04    0.03    0.04    0.00    
Test time         0.01    0.01    0.01    0.00    


***K = 3***

In [ ]:
results3 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=3, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
MSE (testset)     2.2778  2.2995  2.1291  2.2355  0.0757  
RMSE (testset)    1.5092  1.5164  1.4591  1.4949  0.0255  
Fit time          0.08    0.08    0.08    0.08    0.00    
Test time         0.01    0.01    0.01    0.01    0.00    


***K = 4***

In [ ]:
results4 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=4, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 4 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Mean    Std     
MSE (testset)     2.0538  2.2605  2.3535  2.2894  2.2393  0.1123  
RMSE (testset)    1.4331  1.5035  1.5341  1.5131  1.4960  0.0379  
Fit time          0.05    0.05    0.05    0.05    0.05    0.00    
Test time         0.01    0.00    0.00    0.01    0.01    0.00    


***K = 5***

In [ ]:
results5 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=5, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
MSE (testset)     2.2733  2.1968  2.3512  2.2040  2.0999  2.2250  0.0838  
RMSE (testset)    1.5077  1.4822  1.5334  1.4846  1.4491  1.4914  0.0281  
Fit time          0.06    0.05    0.06    0.07    0.06    0.06    0.00    
Test time         0.00    0.00    0.00    0.00    0.00    0.00    0.00    


***K = 6***

In [ ]:
results6 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=6, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 6 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Mean    Std     
MSE (testset)     2.3304  2.1606  2.2055  2.1094  2.2131  2.2666  2.2143  0.0709  
RMSE (testset)    1.5266  1.4699  1.4851  1.4524  1.4876  1.5055  1.4878  0.0238  
Fit time          0.06    0.06    0.06    0.06    0.06    0.06    0.06    0.00    
Test time         0.00    0.01    0.00    0.00    0.00    0.00    0.00    0.00    


***K = 7***

In [ ]:
results7 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=7, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 7 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Fold 7  Mean    Std     
MSE (testset)     1.8619  2.2383  2.2526  2.3449  2.1514  2.1658  2.5666  2.2259  0.1974  
RMSE (testset)    1.3645  1.4961  1.5009  1.5313  1.4668  1.4717  1.6021  1.4905  0.0666  
Fit time          0.06    0.06    0.06    0.06    0.06    0.07    0.06    0.06    0.00    
Test time         0.00    0.00    0.00    0.00    0.00    0.01    0.00    0.00    0.00    


***K = 8***

In [ ]:
results8 = cross_validate(algo, data, measures=['MSE', 'RMSE'], cv=8, verbose=True)

Evaluating MSE, RMSE of algorithm SVD on 8 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Fold 7  Fold 8  Mean    Std     
MSE (testset)     1.8981  2.2572  2.4588  2.6004  1.9783  2.4537  2.1080  1.9937  2.2185  0.2458  
RMSE (testset)    1.3777  1.5024  1.5681  1.6126  1.4065  1.5664  1.4519  1.4120  1.4872  0.0823  
Fit time          0.12    0.10    0.11    0.10    0.11    0.11    0.11    0.11    0.11    0.01    
Test time         0.00    0.01    0.00    0.01    0.00    0.00    0.01    0.00    0.01    0.00    


Berdasarkan hasil *crossfold validation*, didapati bahwa dengan Fold 6 (K=6) memberikan performa yang paling baik dengan rata-rata RMSE sebesar 1.4878.

# **TOP-N**

In [ ]:
# Fungsi untuk memberikan rekomendasi Top-N
def top_n(user_id, n=10):
    # List item
    all_items = list(trainset.all_items())                    # Semua item
    rated_items = [item for item, _ in trainset.ur[user_id]]  # Item yang sudah diberi rating
    unrated_items = list(set(all_items) - set(rated_items))   # Item yang belum diberi rating

    # Prediksi rating untuk item yang belum dinilai oleh pengguna
    predictions = [algo.predict(user_id, item) for item in unrated_items]
    predictions.sort(key=lambda x: x.est, reverse=True)

    # Mengambil 10 item teratas
    top_n = predictions[:n]
    author_name = reviews.loc[reviews['AuthorId'] == user_id]['AuthorName'].values[0]

    return top_n, author_name

In [ ]:
# Spesific User (AuthorId = 1634 || AuthorName = Bill Hilbrich)
user_id = 1634
top_n_recommendations, author_name = top_n(user_id, n=10)

print(f"Rekomendasi Top-N untuk Pengguna {user_id} ({author_name}):")
for i, prediction in enumerate(top_n_recommendations, 1):
    recipe_info = recipes.loc[recipes['RecipeId'] == prediction.iid]
    recipe_name = recipe_info['Name'].values[0] if not recipe_info.empty else "N/A"

    print(f"{i}. RecipeId: {prediction.iid}, Name: {recipe_name}")
    print(f"   Prediksi Rating: {prediction.est}")

Rekomendasi Top-N untuk Pengguna 1634 (Bill Hilbrich):
1. RecipeId: 481, Name: Olive Garden Pasta E Fagioli
   Prediksi Rating: 3.7916844182856098
2. RecipeId: 1069, Name: Rock Candy for Kids
   Prediksi Rating: 3.767633462131844
3. RecipeId: 44, Name: Warm Chicken A La King
   Prediksi Rating: 3.7612442622655493
4. RecipeId: 1780, Name: Buttermilk Cornbread
   Prediksi Rating: 3.7545476151872705
5. RecipeId: 436, Name: Alaskan Blueberry Pie
   Prediksi Rating: 3.7410713874998667
6. RecipeId: 517, Name: Greek Stuffed Meatloaf
   Prediksi Rating: 3.6746928195194153
7. RecipeId: 280, Name: Spanish Potato Omelet (Tortilla a la Espanola)
   Prediksi Rating: 3.674544477301742
8. RecipeId: 632, Name: 100% Whole Wheat Blueberry Muffins
   Prediksi Rating: 3.6269511001673727
9. RecipeId: 210, Name: Christmas Snow Punch
   Prediksi Rating: 3.619476991038499
10. RecipeId: 1035, Name: Puppy Chow Snack Mix
   Prediksi Rating: 3.6182279295152604


In [ ]:
# Spesific User (AuthorId = 1634 || AuthorName = Gay Gilmore ckpt)
user_id = 2046
top_n_recommendations, author_name = top_n(user_id, n=10)

print(f"Rekomendasi Top-N untuk Pengguna {user_id} ({author_name}):")
for i, prediction in enumerate(top_n_recommendations, 1):
    recipe_info = recipes.loc[recipes['RecipeId'] == prediction.iid]
    recipe_name = recipe_info['Name'].values[0] if not recipe_info.empty else "N/A"

    print(f"{i}. RecipeId: {prediction.iid}, Name: {recipe_name}")
    print(f"   Prediksi Rating: {prediction.est}")

Rekomendasi Top-N untuk Pengguna 2046 (Gay Gilmore ckpt):
1. RecipeId: 607, Name: Famous Barr's French Onion Soup
   Prediksi Rating: 4.516121746920541
2. RecipeId: 211, Name: Christmas Cranberry Salad
   Prediksi Rating: 4.340918724063982
3. RecipeId: 517, Name: Greek Stuffed Meatloaf
   Prediksi Rating: 4.332965167203083
4. RecipeId: 44, Name: Warm Chicken A La King
   Prediksi Rating: 4.294691905992785
5. RecipeId: 191, Name: Chilies Corn Muffins
   Prediksi Rating: 4.287004809802006
6. RecipeId: 432, Name: Blue Mountain Blueberry Pie
   Prediksi Rating: 4.2849432240837295
7. RecipeId: 324, Name: Cheesy Stuffed Acorn Squash
   Prediksi Rating: 4.277445552720799
8. RecipeId: 476, Name: English Trifle
   Prediksi Rating: 4.274423171002002
9. RecipeId: 445, Name: Remarkable Rhubarb Bites
   Prediksi Rating: 4.264663924137459
10. RecipeId: 155, Name: Light Cucumber Soup
   Prediksi Rating: 4.256268816808753


In [ ]:
# Spesific User (AuthorId = 26517 || AuthorName = Ann S)
user_id = 26517
top_n_recommendations, author_name = top_n(user_id, n=10)

print(f"Rekomendasi Top-N untuk Pengguna {user_id} ({author_name}):")
for i, prediction in enumerate(top_n_recommendations, 1):
    recipe_info = recipes.loc[recipes['RecipeId'] == prediction.iid]
    recipe_name = recipe_info['Name'].values[0] if not recipe_info.empty else "N/A"

    print(f"{i}. RecipeId: {prediction.iid}, Name: {recipe_name}")
    print(f"   Prediksi Rating: {prediction.est}")

Rekomendasi Top-N untuk Pengguna 26517 (Ann S):
1. RecipeId: 44, Name: Warm Chicken A La King
   Prediksi Rating: 4.191266559200717
2. RecipeId: 1128, Name: Spicy Cherry Glazed Chicken
   Prediksi Rating: 4.168821220805467
3. RecipeId: 432, Name: Blue Mountain Blueberry Pie
   Prediksi Rating: 4.161397181557047
4. RecipeId: 400, Name: Spaetzle - German
   Prediksi Rating: 4.142725228208625
5. RecipeId: 131, Name: Buttermilk-Oat Bread
   Prediksi Rating: 4.124214265151621
6. RecipeId: 445, Name: Remarkable Rhubarb Bites
   Prediksi Rating: 4.1152012127099065
7. RecipeId: 607, Name: Famous Barr's French Onion Soup
   Prediksi Rating: 4.11107702903923
8. RecipeId: 1442, Name: Pig in the Blankets
   Prediksi Rating: 4.109936715989622
9. RecipeId: 476, Name: English Trifle
   Prediksi Rating: 4.099730399572512
10. RecipeId: 1417, Name: Melting Moments
   Prediksi Rating: 4.095733032358499
